Add ethernet support on NUCLEO_H743ZI board

VVESTM opened the pull request:
### Description

This pull request adds ethernet support on STM32 H7 platform.
It is based on the STM32 emac driver. As the H7 HAL is slightly different, I have added ETH_IP_VERSION_V2 switch. This has to be included in targets.json.

Work is based on @CurryGuy job and STM32 cube examples.

Tests done with this command (on IAR, ARM and GCC_ARM toolchains) :
    mbed test -t IAR -m NUCLEO_H743ZI -v -n tests*netsocket*


### Pull request type
    [ ] Fix
    [ ] Refactor
    [X] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@jeromecoutant @LMESTM

VVESTM committed:
Add EMAC support for NUCLEO_H743ZI

This port is based on :
    * CurryGuy ethernet branch :
        https://github.com/CurryGuy/mbed-os/tree/feature-stm32h7-emac
    * STM32 Cube example :
        Applications/LwIP/LwIP_HTTP_Server_Netconn_RTOS example

Signed-off-by: Vincent Veron <vincent.veron@st.com>

VVESTM committed:
lwip: STM32H7: specify heap location

Signed-off-by: Vincent Veron <vincent.veron@st.com>

VVESTM committed:
STM32H7 : update linker script files to use right location for lwip_ram_heap

Signed-off-by: Vincent Veron <vincent.veron@st.com>

jeromecoutant commented on targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H743xI/device/TOOLCHAIN_GCC_ARM/STM32H743xI.ld:
```diff
@@ -180,4 +184,19 @@ SECTIONS
+        *(.ethusbram) 
+
+    } >RAM_D2 AT> FLASH
+}
```
Maybe correct this ?

VVESTM replied:
done

jeromecoutant commented on features/netsocket/emac-drivers/TARGET_STM_EMAC/stm32xx_emac.cpp:
```diff
@@ -475,6 +817,8 @@ void mbed_default_mac_address(char *mac)
 #elif defined (TARGET_STM32F7)
     uint32_t word0 = *(uint32_t *)0x1FF0F420;
+#elif defined (TARGET_STM32H7)
+    uint32_t word0 = *(uint32_t *)0x1FF0F420;
```
Value to check

VVESTM replied:
Replaced with correct value (0x1FF1E800)

0xc0170 commented on targets/targets.json:
```diff
@@ -3221,9 +3222,13 @@
             "EXTRA_IDLE_STACK_REQUIRED",
-            "MBED_TICKLESS"
+            "MBED_TICKLESS",
+            "ETH_IP_VERSION_V2"
```
cant find this used anywhere in this PR?

VVESTM replied:
This is used in features/netsocket/emac-drivers/TARGET_STM_EMAC/stm32xx_emac.cpp file. Used to differentiate H7 specific parts.

jeromecoutant replied:
Because github doesn't expand all files by default :-)

0xc0170 replied:
I've missed that file ! I can see now. Wouldn't the config be better for this or ?

VVESTM replied:
You prefer to add this compilation switch in an other file ? Which one do you have in mind ? (features\netsocket\emac-drivers\TARGET_STM_EMAC\mbed_lib.json for example ?)

I need this switch to be activated for only H7 platform for now. Then, if this new hal is deployed on other platform, we will need to activate it also for those new platforms.

kjbracey replied:
This location seems correct to me - the target descriptor passing information to its drivers about which mode to select. (As long as the option is describing something concrete about the target, that is conceptually not locked to any particular driver implementation - so we mean this target has "V2 hardware" or "V2 HAL").

LMESTM replied:
you may move the stm32xx_emac_config.h file to the family specific folder and add this option there. this way you could tune the IP version per family (and later the thread stack size if needed)

VVESTM replied:
Code updated with @LMESTM solution. Effectively, user is not supposed to change this setting so the .json file can be misunderstanding.
Thanks all for comments and propositions !

VVESTM committed:
Move stm32xx_emac_config.h to family specific folder

This allows to specify which hal version to use for each family.
It can also be used to modify the thread stack size.

Signed-off-by: Vincent Veron <vincent.veron@st.com>
